# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'/Users/yehudahecht/Google Drive/Springboard/Bootcamp/6 - Applying the Data Science Method/6.5 - Modeling'

In [3]:
path = '../data/'
file = 'step3_output.csv'
df = pd.read_csv(path+file, index_col=0)
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0        0.0          0          2     2  ...            1.0   
1        0.0          0          0     0  ...            2.0   
2        0.0          0          0     0  ...            1.0   
3        0.0          1          0     2  ...            2.0   
4        0.0          0          1     2  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 26 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 330 entries, 0 to 329
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   state              330 non-null    object 
 2   summit_elev        330 non-null    int64  
 3   vertical_drop      330 non-null    int64  
 4   base_elev          330 non-null    int64  
 5   trams              330 non-null    int64  
 6   fastEight          330 non-null    float64
 7   fastSixes          330 non-null    int64  
 8   fastQuads          330 non-null    int64  
 9   quad               330 non-null    int64  
 10  triple             330 non-null    int64  
 11  double             330 non-null    int64  
 12  surface            330 non-null    int64  
 13  total_chairs       330 non-null    int64  
 14  Runs               330 non-null    float64
 15  TerrainParks       330 non-null    float64
 16  LongestRun_mi      330 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
# for categorical in ['state','clusters']:
#     states = pd.get_dummies(df[categorical])
#     df[states.columns] = states
#     df.drop(categorical, axis=1, inplace=True)
# df.info()    




states = pd.get_dummies(df['state'])
df[states.columns] = states
df.drop('state', axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 330 entries, 0 to 329
Data columns (total 60 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   summit_elev        330 non-null    int64  
 2   vertical_drop      330 non-null    int64  
 3   base_elev          330 non-null    int64  
 4   trams              330 non-null    int64  
 5   fastEight          330 non-null    float64
 6   fastSixes          330 non-null    int64  
 7   fastQuads          330 non-null    int64  
 8   quad               330 non-null    int64  
 9   triple             330 non-null    int64  
 10  double             330 non-null    int64  
 11  surface            330 non-null    int64  
 12  total_chairs       330 non-null    int64  
 13  Runs               330 non-null    float64
 14  TerrainParks       330 non-null    float64
 15  LongestRun_mi      330 non-null    float64
 16  SkiableTerrain_ac  330 non

In [6]:
pd.set_option('display.max_rows', None)
df.agg([min, max]).T

min                            max
Name               49 Degrees North  Yosemite Ski & Snowboard Area
summit_elev                     315                          13487
vertical_drop                    60                           4425
base_elev                        70                          10800
trams                             0                              4
fastEight                         0                              1
fastSixes                         0                              6
fastQuads                         0                             15
quad                              0                              8
triple                            0                              8
double                            0                             14
surface                           0                             15
total_chairs                      0                             41
Runs                              3                            341
TerrainParks                      0                             14
LongestRun_mi                     0                              6
SkiableTerrain_ac                 8                          26819
Snow Making_ac                    2                           3379
daysOpenLastYear                  3                            305
yearsOpen                         6                           2019
averageSnowfall                  18                            669
AdultWeekday                     15                            179
AdultWeekend                     17                            179
projectedDaysOpen                30                            305
NightSkiing_ac                    0                            650
Alaska                            0                              1
Arizona                           0                              1
California                        0                              1
Colorado                          0                              1
Connecticut                       0                              1
Idaho                             0                              1
Illinois                          0                              1
Indiana                           0                              1
Iowa                              0                              1
Maine                             0                              1
Maryland                          0                              1
Massachusetts                     0                              1
Michigan                          0                              1
Minnesota                         0                              1
Missouri                          0                              1
Montana                           0                              1
Nevada                            0                              1
New Hampshire                     0                              1
New Jersey                        0                              1
New Mexico                        0                              1
New York                          0                              1
North Carolina                    0                              1
Ohio                              0                              1
Oregon                            0                              1
Pennsylvania                      0                              1
Rhode Island                      0                              1
South Dakota                      0                              1
Tennessee                         0                              1
Utah                              0                              1
Vermont                           0                              1
Virginia                          0                              1
Washington                        0                              1
West Virginia                     0                              1
Wisconsin                         0                              1
Wyoming     

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 


/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:14: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
print(y)

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

[ 85.          53.          34.          89.          78.
  47.          64.16681004  64.16681004  64.16681004  78.
  75.          64.16681004  64.16681004  64.16681004 159.
  59.          84.          69.          64.16681004  64.16681004
  83.          64.16681004  79.          50.         125.
  69.          85.         179.          64.16681004  64.16681004
 158.          64.16681004  89.          64.16681004  25.
  64.16681004  79.          89.          71.          79.
  56.          84.          64.16681004  65.         139.
  64.16681004  64.16681004  72.          65.          60.
  55.          62.          42.          64.          70.
  42.          47.          32.          47.          64.16681004
  81.          62.          43.          64.16681004  71.
  55.          64.16681004  35.          40.          45.
  52.          39.          40.          46.          30.
  43.          55.          49.          64.16681004  32.
  75.          99.         105.          79.    

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test, y_pred)

0.934943830806142

In [12]:
explained_variance_score?

In [13]:
y_train_pred = model.predict(X_train)
explained_variance_score(y_train, y_train_pred)

0.9119338128199863

In [14]:
mean_absolute_error(y_test, y_pred)

5.066784156515915

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [15]:
lm.intercept_

64.10401181767517

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [16]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
total_chairs       1.391993e+12
fastQuads          5.277077e+11
surface            4.944226e+11
double             4.357035e+11
triple             3.886775e+11
quad               3.150087e+11
New York           1.698896e+11
Michigan           1.603297e+11
fastSixes          1.564392e+11
Colorado           1.412597e+11
California         1.382358e+11
trams              1.344169e+11
Pennsylvania       1.319133e+11
Wisconsin          1.216344e+11
New Hampshire      1.216344e+11
Vermont            1.179594e+11
Minnesota          1.141403e+11
Utah               1.101623e+11
Montana            1.060073e+11
Idaho              1.060073e+11
Massachusetts      1.016537e+11
Oregon             9.707482e+10
Washington         9.707482e+10
New Mexico         9.223705e+10
Maine              9.223705e+10
Wyoming            8.709727e+10
North Carolina     7.566234e+10
Ohio               6.917645e+10
Connecticut        6.917645e+10
Nevada             6.196842e+10
Illinois           6.196842e+10
Virginia           6.196842e+10
West Virginia      6.196842e+10
Iowa               5.374847e+10
Alaska             5.374847e+10
South Dakota       4.395250e+10
Arizona            4.395250e+10
Indiana            4.395250e+10
Missouri           4.395250e+10
New Jersey         4.395250e+10
Rhode Island       3.112645e+10
Maryland           3.112645e+10
Tennessee          3.112645e+10
fastEight          1.321450e+10
AdultWeekday       1.931817e+01
summit_elev        1.793423e+01
base_elev          1.737383e+01
vertical_drop      5.712340e+00
Runs               1.154793e+00
NightSkiing_ac     8.005369e-01
daysOpenLastYear   4.022797e-01
LongestRun_mi      3.943369e-01
yearsOpen          3.763055e-01
averageSnowfall    3.086208e-01
Snow Making_ac     2.237884e-01
SkiableTerrain_ac  1.800893e-01
projectedDaysOpen  1.621494e-01
TerrainParks       1.287740e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [17]:
to_drop = list(states.columns)
[to_drop.append(i) for i in ['Name','AdultWeekend']]
no_states = df.drop(to_drop, axis=1)
scaler2 = preprocessing.StandardScaler().fit(no_states)
X_scaled = scaler2.transform(no_states)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

model2 = lm.fit(X_train, y_train)
y_pred = model2.predict(X_test)
explained_variance_score(y_test, y_pred)

/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  """


0.924327557929588

In [18]:
mean_absolute_error(y_test, y_pred)

5.501951502893553

In [19]:
pd.DataFrame(abs(lm.coef_), no_states.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         20.236082
summit_elev          12.589033
base_elev             9.767936
vertical_drop         4.353867
averageSnowfall       1.931107
quad                  1.556483
triple                1.386797
surface               1.258499
daysOpenLastYear      1.174548
Runs                  1.061541
fastQuads             0.727970
total_chairs          0.689709
projectedDaysOpen     0.475856
NightSkiing_ac        0.394906
fastSixes             0.370696
double                0.308783
yearsOpen             0.286602
SkiableTerrain_ac     0.269349
LongestRun_mi         0.217660
TerrainParks          0.189711
trams                 0.169124
fastEight             0.073687
Snow Making_ac        0.027834

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [20]:
[to_drop.append(i) for i in ['summit_elev','base_elev']]
no_states_elev = df.drop(to_drop, axis=1)
scaler3 = preprocessing.StandardScaler().fit(no_states_elev)
X_scaled = scaler3.transform(no_states_elev)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

model3 = lm.fit(X_train, y_train)
y_pred = model3.predict(X_test)
explained_variance_score(y_test, y_pred)

/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  after removing the cwd from sys.path.


0.9257250741371149

In [21]:
mean_absolute_error(y_test, y_pred)

5.463450934547313

In [22]:
pd.DataFrame(abs(lm.coef_), no_states_elev.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         20.159761
averageSnowfall       2.264604
quad                  1.597019
triple                1.354509
Runs                  1.253202
surface               1.244158
vertical_drop         1.117666
daysOpenLastYear      1.050756
fastQuads             0.812179
total_chairs          0.733674
NightSkiing_ac        0.531579
fastSixes             0.376235
SkiableTerrain_ac     0.351879
double                0.316856
yearsOpen             0.306203
projectedDaysOpen     0.251268
LongestRun_mi         0.199760
trams                 0.146300
Snow Making_ac        0.145668
TerrainParks          0.105424
fastEight             0.077582

#### Model 4

In [23]:
to_drop = ['Name','AdultWeekend','summit_elev','base_elev']

no_elev = df.drop(to_drop, axis=1)
scaler4 = preprocessing.StandardScaler().fit(no_elev)
X_scaled = scaler4.transform(no_elev)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

model4 = lm.fit(X_train, y_train)
y_pred = model4.predict(X_test)
explained_variance_score(y_test, y_pred)

/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/yehudahecht/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  """


0.9415452939934688

In [24]:
mean_absolute_error(y_test, y_pred)

4.8970044254431215

In [25]:
pd.DataFrame(abs(lm.coef_), no_elev.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
New York           1.029836e+13
Michigan           9.718851e+12
Colorado           8.562871e+12
California         8.379567e+12
Pennsylvania       7.996311e+12
Wisconsin          7.373225e+12
New Hampshire      7.373225e+12
Vermont            7.150454e+12
Minnesota          6.918951e+12
Utah               6.677809e+12
Montana            6.425943e+12
Idaho              6.425943e+12
Massachusetts      6.162038e+12
Oregon             5.884474e+12
Washington         5.884474e+12
New Mexico         5.591218e+12
Maine              5.591218e+12
Wyoming            5.279656e+12
North Carolina     4.586494e+12
total_chairs       4.275933e+12
Ohio               4.193333e+12
Connecticut        4.193333e+12
Nevada             3.756397e+12
Illinois           3.756397e+12
Virginia           3.756397e+12
West Virginia      3.756397e+12
Iowa               3.258121e+12
Alaska             3.258121e+12
Indiana            2.664309e+12
South Dakota       2.664309e+12
Missouri           2.664309e+12
Arizona            2.664309e+12
New Jersey         2.664309e+12
Rhode Island       1.886821e+12
Maryland           1.886821e+12
Tennessee          1.886821e+12
fastQuads          1.621016e+12
surface            1.518770e+12
double             1.338397e+12
triple             1.193942e+12
quad               9.676457e+11
fastSixes          4.805510e+11
trams              4.129028e+11
fastEight          4.059239e+10
AdultWeekday       1.934959e+01
vertical_drop      1.440658e+00
Runs               1.157305e+00
NightSkiing_ac     7.042745e-01
daysOpenLastYear   6.503956e-01
Snow Making_ac     4.920711e-01
LongestRun_mi      4.078444e-01
yearsOpen          3.742967e-01
SkiableTerrain_ac  2.080603e-01
projectedDaysOpen  1.375533e-01
TerrainParks       8.058350e-02
averageSnowfall    6.816861e-02

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.93 | 5.07 | - |
| Model 2. | 0.92 | 5.50 | 'state' |
| Model 3. | 0.93 | 5.46 | 'state','summit_elev','base_elev' |
| Model 4. | 0.94 | 4.90 | 'summit_elev','base_elev' |

Model Selection:

**Model 4**